### PR TITLE
ARROW-9837: [Rust][DataFusion] Add provider for variable

### PR DESCRIPTION
**Pull request by wqc200**
Select @@version;

@@version is a variable, and if we want to get its value, we should get it from outside the system,

### CODE REVIEW

File: rust/datafusion/src/variable/system.rs [hunk 1]
```diff
@@ -0,0 +1,18 @@
...
+
+impl SystemVariable {
+    pub fn new() -> Self {
+        Self {}
+    }
+}
+
+impl VariableProvider for SystemVariable {
+    fn get_value(&self, variable_names: Vec<String>) -> Result<ScalarValue> {
+        let s = format!("{}-{}", "test".to_string(), variable_names.concat());
```
Comment (jorgecarleitao) [inline rust/datafusion/src/variable/system.rs:15]: why `test`?

File: rust/datafusion/src/logicalplan.rs [hunk 2]
```diff
@@ -713,6 +718,9 @@ impl fmt::Debug for Expr {
         match self {
             Expr::Alias(expr, alias) => write!(f, "{:?} AS {}", expr, alias),
             Expr::Column(name) => write!(f, "#{}", name),
+            Expr::ScalarVariable(variable_names) => {
+                write!(f, "#{}", variable_names.join("."))
```
Comment (jorgecarleitao) [inline rust/datafusion/src/logicalplan.rs:722]: this naming is already being used by columns (one line above). Doesn't this make it confusing?

**Comment by andygrove** [comment]
Hi @wqc200 please rebase against master rather than merge from master once this is ready for review.

File: rust/datafusion/src/variable/system.rs (hunk 1 above)
Comment (wqc200) [reply rust/datafusion/src/variable/system.rs]: When the user USES, to create a new system and user_defined method, here is only for the user to understand.
The @@ variable is defined by the external system

File: rust/datafusion/src/logicalplan.rs (hunk 2 above)
Comment (wqc200) [reply rust/datafusion/src/logicalplan.rs]: Yes, i remove the code '#', the variable name itself start with an @ sign

File: rust/datafusion/src/variable/user_defined.rs [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+// under the License.
+
+//! User defined variable provider
+
+use crate::error::Result;
+use crate::logicalplan::ScalarValue;
+use crate::variable::VarProvider;
+
+/// user defined variable
+pub struct UserDefinedVar {}
```
Comment (andygrove) [inline rust/datafusion/src/variable/user_defined.rs:25]: This seems to be identical to `SystemVar`. I understand that we want to register separate variables for user-defined and system, but I don't think we need two separate implementations of the trait?

File: rust/datafusion/src/variable/system.rs [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+    /// new system variable
+    pub fn new() -> Self {
+        Self {}
+    }
+}
+
+impl VarProvider for SystemVar {
+    /// get system variable value
+    fn get_value(&self, var_names: Vec<String>) -> Result<ScalarValue> {
+        let s = format!("{}-{}", "test".to_string(), var_names.concat());
```
Comment (andygrove) [inline rust/datafusion/src/variable/system.rs:37]: This seems to be test code but is not marked with `#[test]` as far as I can see.

File: rust/datafusion/src/variable/user_defined.rs (hunk 3 above)
Comment (wqc200) [reply rust/datafusion/src/variable/user_defined.rs]: Yes, I move it to test dir

File: rust/datafusion/src/variable/system.rs (hunk 4 above)
Comment (wqc200) [reply rust/datafusion/src/variable/system.rs]: Yes, I move it to test dir

**Review by andygrove — APPROVED** [review]
LGTM. Thanks @wqc200

**Comment by andygrove** [comment]
@wqc200 Thanks for making those changes. However, I still see merges in the commit history for this PR. I would suggest creating a new PR from master with these approved changes if you are not comfortable with rebasing this PR. 

See this article for more information about rebasing: https://andygrove.io/apache_arrow_git_tips/

**Pull request closed without merging**